pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1366,5 +1388,80 @@ public void markBatchMessagePublished() {
         this.hasBatchMessagePublished = true;
     }
 
+    public RateLimiter getDispatchRateLimiter() {
+        return dispatchRateLimiter;
+    }
+    
+    public DispatchRateType getDispatchRateType() {
+        return dispatchRateType;
+    }
+
+    public boolean tryDispatchPermit(long permits) {
+        return permits <= 0 || dispatchRateLimiter == null
+        // acquiring permits must be < configured msg-rate;
+                || dispatchRateLimiter.tryAcquire(permits);
+    }
+    
+    public boolean hasMessageDispatchPermit() {
+        return dispatchRateLimiter == null || dispatchRateLimiter.getAvailablePermits() > 0;
+    }
+
+    private void updateDispatchRate(LocalZooKeeperCacheService localZKCache) {
+        final String path = joinPath(LOCAL_POLICIES_ROOT, DestinationName.get(this.topic).getNamespace());
+        try {
+            Optional<LocalPolicies> policies = localZKCache.policiesCache().getAsync(path).get(cacheTimeOutInSec,
+                    SECONDS);
+            if (policies.isPresent() && policies.get().dispatchRate != null
+                    && policies.get().dispatchRate.topicDispatchRate > 0) {
+                updateDispatchRate(policies.get().dispatchRate);
+                log.info("[{}] configured message-dispatch rate configured at policy {}}", this.topic,
+                        policies.get().dispatchRate);
+                return;
+            }
+        } catch (Exception e) {
+            log.warn("Failed to get message-rate for {}", this.topic, e);
+        }
+        DispatchRate dispatchRate = brokerService.pulsar().getConfiguration().getDispatchRatePerTopicInMsg() > 0
```

Comment:
It would require 2 rate limiters only if both limits are set. Otherwise one of them, or both, will be null.